vm/src/dictdatatype.rs

```diff
@@ -273,23 +274,63 @@ enum LookupResult {
     Existing(EntryIndex), // Existing record, index into entries
 }
 
-#[cfg_attr(feature = "flame-it", flame())]
-fn collection_hash(vm: &VirtualMachine, object: &PyObjectRef) -> PyResult<HashValue> {
-    let raw_hash = vm._hash(object)?;
-    let mut hasher = DefaultHasher::new();
-    raw_hash.hash(&mut hasher);
-    Ok(hasher.finish() as HashValue)
+/// Types implementing this trait can be used to index
+/// the dictionary. Typical usecases are:
+/// - PyObjectRef -> arbitrary python type used as key
+/// - str -> string reference used as key, this is often used internally
+pub trait DictKey {
+    fn do_hash(&self, vm: &VirtualMachine) -> PyResult<HashValue>;
+    fn do_is(&self, other: &PyObjectRef) -> bool;
+    fn do_eq(&self, vm: &VirtualMachine, other_key: &PyObjectRef) -> PyResult<bool>;
 }
 
-/// Invoke __eq__ on two keys
-fn do_eq(vm: &VirtualMachine, key1: &PyObjectRef, key2: &PyObjectRef) -> Result<bool, PyObjectRef> {
-    let result = vm._eq(key1.clone(), key2.clone())?;
-    objbool::boolval(vm, result)
+/// Implement trait for PyObjectRef such that we can use python objects
+/// to index dictionaries.
+impl DictKey for PyObjectRef {
+    fn do_hash(&self, vm: &VirtualMachine) -> PyResult<HashValue> {
+        let raw_hash = vm._hash(self)?;
+        let mut hasher = DefaultHasher::new();
+        raw_hash.hash(&mut hasher);
+        Ok(hasher.finish() as HashValue)
+    }
+
+    fn do_is(&self, other: &PyObjectRef) -> bool {
+        self.is(other)
+    }
+
+    fn do_eq(&self, vm: &VirtualMachine, other_key: &PyObjectRef) -> PyResult<bool> {
+        let result = vm._eq(self.clone(), other_key.clone())?;
+        objbool::boolval(vm, result)
+    }
+}
+
+/// Implement trait for the str type, so that we can use strings
+/// to index dictionaries.
+impl DictKey for String {
+    fn do_hash(&self, _vm: &VirtualMachine) -> PyResult<HashValue> {
+        let raw_hash = pyhash::hash_value(self).to_bigint().unwrap();
+        let raw_hash = pyhash::hash_bigint(&raw_hash);
+        let mut hasher = DefaultHasher::new();
+        raw_hash.hash(&mut hasher);
+        Ok(hasher.finish() as HashValue)
+    }
+
+    fn do_is(&self, _other: &PyObjectRef) -> bool {
+        // No matter who the other pyobject is, we are never the same thing, since
+        // we are a str, not a pyobject.
+        false
+    }
+
+    fn do_eq(&self, vm: &VirtualMachine, other_key: &PyObjectRef) -> PyResult<bool> {
+        // Fall back to PyString implementation.
+        let s = vm.new_str(self.to_string());
+        s.do_eq(vm, other_key)
```

Comment:
I feel like we could have a more efficient eq implementation here than reallocating the string.

Reply:
Absolutely, this involves checking the specific python type to be of `PyStringRef`, and then doing a string compare. I suspect this happens only in a couple of cases, after the first hash compare is a success. I will implement this though, since I think it is a good idea.